core/src/main/java/org/elasticsearch/index/query/MatchQueryBuilder.java

```diff
@@ -123,9 +110,13 @@ public MatchQueryBuilder(StreamInput in) throws IOException {
         super(in);
         fieldName = in.readString();
         value = in.readGenericValue();
-        type = MatchQuery.Type.readFromStream(in);
+        if (in.getVersion().onOrAfter(Version.V_6_0_0_rc1)) {
```

Comment:
This assumes these changes should and can be backported to 6.0.0

Reply:
The logic seems wrong. I think it should be `before` and not `onOrAfter` ? 
Also you should do that in multiple steps to avoid build failures. Start with current version (v7), then backport to v6 and when the build is green you can change the expected version here to be v6. You can add a TODO comment to make sure that reviewers understand the expectations.

Reply:
Thanks, of course it should be before, bad copy error on my side. Looking at this, I think I will pull out the MatchQueryBuilder changes to a separate PR since I think it will make backporting easier, the rest of this PR is more straight forward.